src/partr.c

```diff
@@ -405,7 +402,10 @@ static jl_task_t *get_next_task(jl_value_t *trypoptask, jl_value_t *q)
 
 static int may_sleep(jl_ptls_t ptls)
 {
-    return ptls->sleep_check_state == sleeping;
+    // sleep_check_state is only transitioned from not_sleeping to sleeping
+    // by the thread itself. As a result, if this returns false, it will
+    // continue returning false. If it returns true, there are no guarantees.
+    return jl_atomic_load_relaxed(&ptls->sleep_check_state) == sleeping;
```

Comment:
Do we need something stronger than relaxed ordering here, because `sleep_check_state` is meant to ensure consistency of other loads/stores?

Reply:
The actual condition variable is protected by a lock, so we don't depend on any memory access dependencies here, I don't think. That's why I picked relaxed. Do you disagree?

Reply:
> Do you disagree?

I don't disagree, I'm just unsure.

The condition variable seems fine. But I noticed that `may_sleep` is also used in a branch related to the event loop which is protected with `jl_uv_mutex`.  So `may_sleep` doesn't always go along with the same lock. Which seems potentially confusing but perhaps harmless.